File: sub_pub/metrics/collector.py
"""Metrics collection for monitoring message flow"""
from dataclasses import dataclass, field
from typing import Dict
from threading import Lock
import time


@dataclass
class TopicMetrics:
    """Metrics for a specific topic"""
    message_count: int = 0
    total_bytes: int = 0
    error_count: int = 0
    last_message_time: float = 0.0
    
    def record_message(self, size: int) -> None:
        """Record a successful message"""
        self.message_count += 1
        self.total_bytes += size
        self.last_message_time = time.time()
    
    def record_error(self) -> None:
        """Record an error"""
        self.error_count += 1


class MetricsCollector:
    """Thread-safe metrics collector"""
    
    def __init__(self):
        self._source_metrics: Dict[str, TopicMetrics] = {}
        self._destination_metrics: Dict[str, TopicMetrics] = {}
        self._lock = Lock()
        self._start_time = time.time()
    
    def record_source_message(self, topic: str, size: int) -> None:
        """Record a message received from a source topic"""
        with self._lock:
            if topic not in self._source_metrics:
                self._source_metrics[topic] = TopicMetrics()
            self._source_metrics[topic].record_message(size)
    
    def record_destination_message(self, topic: str, size: int) -> None:
        """Record a message sent to a destination topic"""
        with self._lock:
            if topic not in self._destination_metrics:
                self._destination_metrics[topic] = TopicMetrics()
            self._destination_metrics[topic].record_message(size)
    
    def record_source_error(self, topic: str) -> None:
        """Record an error from a source topic"""
        with self._lock:
            if topic not in self._source_metrics:
                self._source_metrics[topic] = TopicMetrics()
            self._source_metrics[topic].record_error()
    
    def record_destination_error(self, topic: str) -> None:
        """Record an error to a destination topic"""
        with self._lock:
            if topic not in self._destination_metrics:
                self._destination_metrics[topic] = TopicMetrics()
            self._destination_metrics[topic].record_error()
    
    def get_metrics(self) -> Dict:
        """Get all metrics"""
        with self._lock:
            uptime = time.time() - self._start_time
            return {
                'uptime_seconds': uptime,
                'source_metrics': {
                    topic: {
                        'message_count': metrics.message_count,
                        'total_bytes': metrics.total_bytes,
                        'error_count': metrics.error_count,
                        'last_message_time': metrics.last_message_time,
                        'rate_per_second': metrics.message_count / uptime if uptime > 0 else 0
                    }
                    for topic, metrics in self._source_metrics.items()
                },
                'destination_metrics': {
                    topic: {
                        'message_count': metrics.message_count,
                        'total_bytes': metrics.total_bytes,
                        'error_count': metrics.error_count,
                        'last_message_time': metrics.last_message_time,
                        'rate_per_second': metrics.message_count / uptime if uptime > 0 else 0
                    }
                    for topic, metrics in self._destination_metrics.items()
                }
            }
    
    def reset(self) -> None:
        """Reset all metrics"""
        with self._lock:
            self._source_metrics.clear()
            self._destination_metrics.clear()
            self._start_time = time.time()

File: rust/src/metrics/collector.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use serde::Serialize;

/// Per-topic statistics.
#[derive(Debug, Default, Clone, Serialize)]
pub struct TopicMetrics {
    pub message_count: u64,
    pub total_bytes: u64,
    pub error_count: u64,
    pub last_message_time: Option<std::time::SystemTime>,
}

/// Snapshot of all metrics at a point in time.
#[derive(Debug, Serialize)]
pub struct MetricsSnapshot {
    pub uptime_secs: f64,
    pub source_topics: HashMap<String, TopicMetrics>,
    pub destination_topics: HashMap<String, TopicMetrics>,
    pub total_messages_received: u64,
    pub total_messages_sent: u64,
    pub total_errors: u64,
    /// Overall message throughput (messages per second since start).
    pub messages_per_second: f64,
}

/// Thread-safe metrics collector.
///
/// All methods take `&self` (not `&mut self`) so a single `Arc<MetricsCollector>`
/// can be shared freely across tasks and threads.
#[derive(Debug)]
pub struct MetricsCollector {
    started_at: Instant,
    inner: Arc<Mutex<Inner>>,
}

#[derive(Debug, Default)]
struct Inner {
    source_topics: HashMap<String, TopicMetrics>,
    destination_topics: HashMap<String, TopicMetrics>,
}

impl MetricsCollector {
    /// Create a new collector.
    pub fn new() -> Self {
        Self {
            started_at: Instant::now(),
            inner: Arc::new(Mutex::new(Inner::default())),
        }
    }

    /// Record a message received from a source topic.
    pub fn record_source_message(&self, topic: &str, size: usize) {
        let mut guard = self.inner.lock().expect("metrics lock poisoned");
        let entry = guard.source_topics.entry(topic.to_string()).or_default();
        entry.message_count += 1;
        entry.total_bytes += size as u64;
        entry.last_message_time = Some(std::time::SystemTime::now());
    }

    /// Record a message successfully published to a destination topic.
    pub fn record_destination_message(&self, topic: &str, size: usize) {
        let mut guard = self.inner.lock().expect("metrics lock poisoned");
        let entry = guard
            .destination_topics
            .entry(topic.to_string())
            .or_default();
        entry.message_count += 1;
        entry.total_bytes += size as u64;
        entry.last_message_time = Some(std::time::SystemTime::now());
    }

    /// Record an error on a source topic.
    pub fn record_source_error(&self, topic: &str) {
        let mut guard = self.inner.lock().expect("metrics lock poisoned");
        guard
            .source_topics
            .entry(topic.to_string())
            .or_default()
            .error_count += 1;
    }

    /// Record an error on a destination topic.
    pub fn record_destination_error(&self, topic: &str) {
        let mut guard = self.inner.lock().expect("metrics lock poisoned");
        guard
            .destination_topics
            .entry(topic.to_string())
            .or_default()
            .error_count += 1;
    }

    /// Build a point-in-time snapshot of all metrics.
    pub fn snapshot(&self) -> MetricsSnapshot {
        let guard = self.inner.lock().expect("metrics lock poisoned");
        let uptime = self.started_at.elapsed().as_secs_f64();

        let total_received: u64 = guard.source_topics.values().map(|t| t.message_count).sum();
        let total_sent: u64 = guard
            .destination_topics
            .values()
            .map(|t| t.message_count)
            .sum();
        let total_errors: u64 = guard
            .source_topics
            .values()
            .chain(guard.destination_topics.values())
            .map(|t| t.error_count)
            .sum();

        let msgs_per_sec = if uptime > 0.0 {
            total_received as f64 / uptime
        } else {
            0.0
        };

        MetricsSnapshot {
            uptime_secs: uptime,
            source_topics: guard.source_topics.clone(),
            destination_topics: guard.destination_topics.clone(),
            total_messages_received: total_received,
            total_messages_sent: total_sent,
            total_errors,
            messages_per_second: msgs_per_sec,
        }
    }
}

impl Default for MetricsCollector {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn records_source_messages() {
        let m = MetricsCollector::new();
        m.record_source_message("topic-a", 100);
        m.record_source_message("topic-a", 200);
        let snap = m.snapshot();
        let t = &snap.source_topics["topic-a"];
        assert_eq!(t.message_count, 2);
        assert_eq!(t.total_bytes, 300);
    }

    #[test]
    fn records_destination_messages() {
        let m = MetricsCollector::new();
        m.record_destination_message("out", 50);
        let snap = m.snapshot();
        assert_eq!(snap.total_messages_sent, 1);
    }

    #[test]
    fn records_errors() {
        let m = MetricsCollector::new();
        m.record_source_error("topic-x");
        m.record_source_error("topic-x");
        let snap = m.snapshot();
        assert_eq!(snap.total_errors, 2);
    }
}
